root: expose CONN_MAX_AGE, CONN_HEALTH_CHECKS and DISABLE_SERVER_SIDE_CURSORS for PostgreSQL config

tomasfarias opened the pull request:
<!--
👋 Hi there! Welcome.

Please check the Contributing guidelines: https://goauthentik.io/developer-docs/#how-can-i-contribute
-->

## Details

<!--
Explain what this PR changes, what the rationale behind the change is, if any new requirements are introduced or any breaking changes caused by this PR.

Ideally also link an Issue for context that this PR will close using `closes #`
-->
Closes #9152
Closes #12278

Authentik enables unlimited persistent connections (`CONN_MAX_AGE=None`) when setting `USE_PGBOUNCER=True`. However, unlimited persistent connections can cause problems when running with default PgBouncer configuration: In particular, it would seem that the PgBouncer defaults `pool_mode="session"` and `query_wait_timeout=120` cause DB connections to be dropped while Authentik's Django expects them to be up forever. 

To address this, and accommodate more PgBouncer configurations, we can expose the `CONN_MAX_AGE` configuration setting, instead of hardcoding it to `None`. In my own deployment, setting `CONN_MAX_AGE=0` addressed all connection issues. That being said, it would be good to get feedback from other users in #9152 to confirm my findings.

### Changes
* Expose `CONN_MAX_AGE` via a new `PGBOUNCER_CONN_MAX_AGE` config key.
* Document it.

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [ ] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [ ] The code has been formatted (`make web`)

If applicable

-   [x] The documentation has been updated
-   [x] The documentation has been formatted (`make website`)

tomasfarias committed:
authentik/root: Expose CONN_MAX_AGE configuration setting

When running PgBouncer, there is no need to have Django hold unlimited
persistent connections; that's what PgBouncer is doing. Moreover, any
timeouts will cause the connection to be dropped from under Django,
which anecdotically doesn't seem to behave well.

So, we can let users set CONN_MAX_AGE whatever to whatever they need
with a new configuration.

Also updates the docs to reference this.

rissson commented:
> Happy to work on a follow-up commit to also expose `disable_server_side_cursors` and deprecate `use_pgbouncer` (+ doc updates). Unfortunately, I am not familiar with pgpool to also contribute changes for pgpool settings.

With pleasure! Let me know if you need any guidance for anything.

You already know about `authentik/root/settings.py`, but there is also `authentik/root/db/base.py` which may (?) need some changes.

Docs wise, settings are documentation in `website/docs/installation/configuration.mdx`. As for release notes, there is a template you can copy from `website/docs/releases/_template.md` to `website/docs/releases/2024/next.md`.

We can probably get away with first checking if use_pgpool or use_pgbouncer are set, configure conn_max_age and disable_server_side_cursor, and then override those if they are in the configuration as well. The tricky bit is to make it work for read-replicas in an intuitive way.

As for the documenting the options required for pgpool, it'd be as simple as saying that if pg_pool is used, disable_server_side_cursors must be true.

tomasfarias commented:
Just to clarify:

> We can probably get away with first checking if use_pgpool or use_pgbouncer are set, configure conn_max_age and disable_server_side_cursor, and then override those if they are in the configuration as well.

So the idea is to keep both use_pgpool and use_pgbouncer? I thought that we were removing them in favor of exposing only conn_max_age and disable_server_side_cursors. Or are we keeping them in a "marked for deprecation" state to ease with the migration?

Keeping them does make it confusing when working with replicas, as use_pgpool and use_pgbouncer are more like flags indicating which settings to use: So, should we check and accept a flag per replica or the global? By removing use_pgpool and use_pgbouncer, we could let folks set replica overrides for conn_max_age and disable_server_side_cursors like any other db setting.

tomasfarias committed:
authentik/root: Deprecate USE_<pooler> in favor of db settings

The `USE_PGBOUNCER` and `USE_PGPOOL` were used to set hardcoded
django db settings to work with the corresponding pooler. However,
these settings could be incompatible with even default configurations
of poolers (i.e. PgBouncer).

To allow supporting a wider set of pooler configurations, we deprecate
both of the `USE_<pooler>` settings in favor of exposing the underlying db
settings that were changed as these two pooler settings were set:
`CONN_MAX_AGE` and `DISABLE_SERVER_SIDE_CURSORS`.

This is a breaking change as it requries users who had `USE_<pooler>`
set to `true` to now configure `CONN_MAX_AGE` and
`DISABLE_SERVER_SIDE_CURSORS`. The next release notes include the
values to use to maintain functionality, and the documentation
includes a new section with common pitfalls with db connection poolers.

tomasfarias commented:
My latest commit goes ahead with deprecating `USE_<pooler>` settings as it was my original understanding. Granted, I am more than happy to look into bringing them back after my previous comment is clarified, if this was not the intended route. I just happened to have some time now and decided to move forward with my original idea instead of waiting. 

The commit also includes documentation changes and next release notes updates as instructed.

rissson commented on authentik/root/settings.py:
```diff
@@ -305,14 +305,15 @@
-    DATABASES["default"]["DISABLE_SERVER_SIDE_CURSORS"] = True
-    # https://docs.djangoproject.com/en/4.0/ref/databases/#persistent-connections
-    DATABASES["default"]["CONN_MAX_AGE"] = None  # persistent
+# https://docs.djangoproject.com/en/4.0/ref/databases/#transaction-pooling-server-side-cursors
+DATABASES["default"]["DISABLE_SERVER_SIDE_CURSORS"] = CONFIG.get_bool(
+    "postgresql.disable_server_side_cursors", False
+)
+# https://docs.djangoproject.com/en/4.0/ref/databases/#persistent-connections
+# Not a PostgreSQL-specific setting, but a database-specific setting.
+# Only PostgreSQL is supported as a database, so we place this setting it in the 'postgresql' path.
+DATABASES["default"]["CONN_MAX_AGE"] = CONFIG.get("postgresql.conn_max_age", None)
```
Ideally, all of that would directly happen above in the `DATABASES["default"]` dict directly.

rissson commented on authentik/root/settings.py:
```diff
@@ -305,14 +305,15 @@
+# https://docs.djangoproject.com/en/4.0/ref/databases/#persistent-connections
+# Not a PostgreSQL-specific setting, but a database-specific setting.
+# Only PostgreSQL is supported as a database, so we place this setting it in the 'postgresql' path.
+DATABASES["default"]["CONN_MAX_AGE"] = CONFIG.get("postgresql.conn_max_age", None)
```
```suggestion
DATABASES["default"]["CONN_MAX_AGE"] = CONFIG.get("postgresql.conn_max_age", 0)
```
Which is the default. Also, we should use `get_int` here, but we need to figure out a way for that to be able to return `None` if that's what is set by users.

tomasfarias replied:
> Also, we should use `get_int` here, but we need to figure out a way for that to be able to return None if that's what is set by users.

I see a problem:
* Making `get_int` return `None` would require changing the type signature to `int | None` (regardless of how it's implemented). I _think_ that would add too much of a mental load as every call to `get_int` now needs to account for `None`.

For this reason, I would define a new `get_optional_int` with the `-> int | None` signature, and would check for env variables set to the string `None` (case-insensitive). This is similar to how `get_bool` works, and also makes it obvious this function may not return an int.

If you disagree, though, I can move the implementation to `get_int` instead, and hope there aren't many type errors to wrap around a `if not None else default`.

tomasfarias replied:
On another topic: Agree with setting `0` as the default `CONN_MAX_AGE`, as that's the Django actual default, I've made this change too :+1:

rissson replied:
Yeah I agree about the `get_optional_int`. `CONN_MAX_AGE` still defaults to None from what I'm reading though

rissson replied:
> `CONN_MAX_AGE` still defaults to None from what I'm reading though

Only in the documentation, my bad

rissson commented:
> My latest commit goes ahead with deprecating USE_<pooler> settings as it was my original understanding

Yeah I think that's the best route. I was just saying that in theory we could preserve backwards compatibility.

rissson reviewed (changes requested):
Before we go ahead with this, we'd probably want to check out https://docs.djangoproject.com/en/dev/releases/5.1/#postgresql-connection-pools as well, just in case changing defaults might make sense (so not using my suggestions below).

tomasfarias committed:
authentik/root: Support setting CONN_MAX_AGE as None

Support was added in the form of a new `get_optional_int` method for
CONFIG, to avoid requiring all existing uses of `get_int` to account
for `None`.

Moreover, commit moves settings under the `"DATABASES"` dictionary.

tomasfarias commented:
> Before we go ahead with this, we'd probably want to check out [docs.djangoproject.com/en/dev/releases/5.1/#postgresql-connection-pools](https://docs.djangoproject.com/en/dev/releases/5.1/#postgresql-connection-pools) as well, just in case changing defaults might make sense

I've taken a look, and it doesn't seem that Django is changing any settings when enabling the internal pool. Moreover the connection pool appears disabled by default. 

I _think_ that when/if a `AUTHENTIK_POSTGRESQL__POOL` setting is ever exposed, then we would have to document how `CONN_MAX_AGE` and `DISABLE_SERVER_SIDE_CURSORS` should be set too, as well as the pool specific settings.

rissson commented on authentik/lib/default.yml:
```diff
@@ -6,8 +6,8 @@ postgresql:
   password: "env://POSTGRES_PASSWORD"
-  use_pgbouncer: false
-  use_pgpool: false
+  conn_max_age: null
```
```suggestion
  conn_max_age: 0
```

rissson commented on authentik/root/settings.py:
```diff
@@ -302,18 +302,18 @@
-    DATABASES["default"]["DISABLE_SERVER_SIDE_CURSORS"] = True
-    # https://docs.djangoproject.com/en/4.0/ref/databases/#persistent-connections
-    DATABASES["default"]["CONN_MAX_AGE"] = None  # persistent
-
```
We'll have an issue in the replica settings that get the data with `get` instead of `get_optional_int`. We probably need special cases in there.

rissson commented on website/docs/installation/configuration.mdx:
```diff
@@ -70,14 +70,14 @@ To check if your config has been applied correctly, you can run the following co
 -   `AUTHENTIK_POSTGRESQL__SSLROOTCERT`: CA root for server ssl verification
 -   `AUTHENTIK_POSTGRESQL__SSLCERT`: Path to x509 client certificate to authenticate to server
 -   `AUTHENTIK_POSTGRESQL__SSLKEY`: Path to private key of `SSLCERT` certificate
+-   `AUTHENTIK_POSTGRESQL__CONN_MAX_AGE`: Database connection lifetime. Defaults to `null` (unlimited persistent connections).
```
```suggestion
-   `AUTHENTIK_POSTGRESQL__CONN_MAX_AGE`: Database connection lifetime. Defaults to `0` (no persistent connections). Can be set to `none` for unlimited persistent connections.
```

rissson commented on website/docs/installation/configuration.mdx:
```diff
@@ -70,14 +70,14 @@ To check if your config has been applied correctly, you can run the following co
+-   `AUTHENTIK_POSTGRESQL__DISABLE_SERVER_SIDE_CURSORS`: Disable server side cursors when set to `true`. Defaults to `false`.
 
-All PostgreSQL settings, apart from `USE_PGBOUNCER` and `USE_PGPOOL`, support hot-reloading. Adding and removing read replicas doesn't support hot-reloading.
+All PostgreSQL settings support hot-reloading. Adding and removing read replicas doesn't support hot-reloading.
```
This is actually false, we only support `host`, `port`, `user` and `password` for hot-reloading (see `authentik/root/db/base.py`). You can either update the documentation or the code, as you want.

tomasfarias replied:
True, I've missed this. Thanks for pointing it out. Will update docs as it seems like the least impact, but personally wouldn't be against making all hot-reloadable.

rissson commented on website/docs/installation/configuration.mdx:
```diff
@@ -96,8 +96,20 @@ The same PostgreSQL settings as described above are used for each read replica.
-Note that `USE_PGBOUNCER` and `USE_PGPOOL` are inherited from the main database configuration and are _not_ overridable on read replicas.
+### Using a PostgreSQL connection pooler (PgBouncer or PgPool)
+
+When your PostgreSQL database/s are running behind a connection pooler, like PgBouncer or PgPool, two settings may need to be overridden:
```
```suggestion
When your PostgreSQL database/s are running behind a connection pooler, like PgBouncer or PgPool, two settings need to be overridden:
```

rissson commented on website/docs/installation/configuration.mdx:
```diff
@@ -96,8 +96,20 @@ The same PostgreSQL settings as described above are used for each read replica.
+When your PostgreSQL database/s are running behind a connection pooler, like PgBouncer or PgPool, two settings may need to be overridden:
+
+-   `AUTHENTIK_POSTGRESQL__CONN_MAX_AGE`
+    A connection pooler running in session pool mode (PgBouncer default) can be incompatible with unlimited persistent connections enabled by leaving this setting in its default value (`null`): If the connection from the connection pooler to the database server is dropped, the connection pooler will wait for the client to disconnect before releasing the connection; however this will **never** hap ...
```
This needs changing as well, since the default will be `0`.

cardboardpig commented:
Currently seeing the issue fixed by this MR against a Percona Postgres PGBouncer instance; I have set `AUTHENTIK_POSTGRESQL__USE_PGBOUNCER` to false and `AUTHENTIK_POSTGRESQL__USE_PGPOOL` to true as a dirty work around to get the pods to stay up while testing but it's not ideal :sweat_smile:

BeryJu commented:
@tomasfarias Could you look into the changes requested above? Or let us know if you don't have the time/resources to continue with this PR

tomasfarias commented:
@BeryJu Hey! Apologies for the long absence. My day job has been quite demanding, and I had some time-off scheduled to recover.

I'll be taking a look at this PR this weekend.

tomasfarias committed:
authentik/root: Update default to 0

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: Tomás Farías Santana <tomas@tomasfarias.dev>

tomasfarias committed:
authentik/root: Add special case for replica settings

tomasfarias committed:
authentik/root: Update default in documentation

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: Tomás Farías Santana <tomas@tomasfarias.dev>

tomasfarias committed:
authentik/root: Only 4 settings support hot reloading

tomasfarias committed:
authentik/root: Remove "may" in docs.

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: Tomás Farías Santana <tomas@tomasfarias.dev>

tomasfarias committed:
authentik/root: Use 'null' and update paragraph

tomasfarias committed:
authentik/root: Use 'none' instead of 'null'

tomasfarias commented on website/docs/installation/configuration.mdx:
```diff
@@ -70,14 +70,14 @@ To check if your config has been applied correctly, you can run the following co
+-   `AUTHENTIK_POSTGRESQL__DISABLE_SERVER_SIDE_CURSORS`: Disable server side cursors when set to `true`. Defaults to `false`.
 
-All PostgreSQL settings, apart from `USE_PGBOUNCER` and `USE_PGPOOL`, support hot-reloading. Adding and removing read replicas doesn't support hot-reloading.
+The PostgreSQL settings `HOST`, `PORT`, `USER`, and `PASSWORD` support hot-reloading. Adding and removing read replicas doesn't support hot-reloading.
```
This should be more accurate than before, as it's only these four settings that support hot-reloading.

tomasfarias commented:
Afaik all conversations should be addressed so this should be good for another review + hopefully merge?

rissson commented on website/docs/installation/configuration.mdx:
```diff
@@ -70,14 +70,14 @@ To check if your config has been applied correctly, you can run the following co
 -   `AUTHENTIK_POSTGRESQL__SSLROOTCERT`: CA root for server ssl verification
 -   `AUTHENTIK_POSTGRESQL__SSLCERT`: Path to x509 client certificate to authenticate to server
 -   `AUTHENTIK_POSTGRESQL__SSLKEY`: Path to private key of `SSLCERT` certificate
+-   `AUTHENTIK_POSTGRESQL__CONN_MAX_AGE`: Database connection lifetime. Defaults to `0` (no persistent connections). Can be set to `"none"` for unlimited persistent connections.
```
I'm not a huge fan of having a custom value to represent `null`, especially since YAML `null` is valid. But I don't have any specific idea on how to do this anyway.

rissson committed:
website/docs: 2024.12 release notes

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
more release notes

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
add highlight

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
add impersonation breaking change

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

Commit:
add to sidebar.js

rissson committed:
Merge branch 'main' into feat/expose-conn-max-age-for-postgresql

rissson committed:
Merge branch 'docs-2024.12-release-notes' into feat/expose-conn-max-age-for-postgresql

rissson committed:
update release notes

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
temporarly remove from default config

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Merge branch 'main' into feat/expose-conn-max-age-for-postgresql

rissson committed:
fix link

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
fix tests

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson commented on website/docs/releases/2024/v2024.12.md:
```diff
@@ -24,6 +24,16 @@ To try out the release candidate, replace your Docker image tag with the latest
 
     You can disable this behavior in the **Admin interface** under **System** > **Settings**.
 
+- **Deprecated PostgreSQL `USE_PGBOUNCER` and `USE_PGPOOL` settings**
```
Might be for next release actually

rissson commented on website/docs/install-config/configuration/configuration.mdx:
```diff
@@ -96,8 +99,23 @@ The same PostgreSQL settings as described above are used for each read replica.
+When your PostgreSQL database/s are running behind a connection pooler, like PgBouncer or PgPool, two settings need to be overridden:
+
+- `AUTHENTIK_POSTGRESQL__CONN_MAX_AGE`
+  A connection pooler running in session pool mode (PgBouncer default) can be incompatible with unlimited persistent connections enabled by setting this to `"none"`: If the connection from the connection pooler to the database server is dropped, the connection pooler will wait for the client to disconnect before releasing the connection; however this will **never** happen as authentik is configur ...
```
```suggestion
  A connection pooler running in session pool mode (PgBouncer default) can be incompatible with unlimited persistent connections enabled by setting this to `null`: If the connection from the connection pooler to the database server is dropped, the connection pooler will wait for the client to disconnect before releasing the connection; however this will **never** happen as authentik is configured to keep the connection to the connection pooler forever.
```

rissson committed:
Update website/docs/install-config/configuration/configuration.mdx

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
fix link

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
some more formatting

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson commented:
/cherry-pick version-2024.12
